xcode/Subconscious/Shared/Services/NoosphereService.swift

```diff
@@ -0,0 +1,204 @@
+        }
+        
+        guard let sphereMnemonicPointer = ns_sphere_receipt_mnemonic(
+            sphereReceipt
+        ) else {
+            throw NoosphereError.foreignError(
+                "Failed to get pointer for mnemonic"
+            )
+        }
+        defer {
+            ns_string_free(sphereMnemonicPointer)
+        }
+        
+        let sphereIdentity = String.init(cString: sphereIdentityPointer)
+        let sphereMnemonic = String.init(cString: sphereMnemonicPointer)
+        
+        return SphereReceipt(
+            identity: sphereIdentity,
+            mnemonic: sphereMnemonic
+        )
+    }
+
+    /// Perform an action with SphereFS using a closure.
+    /// SphereFS pointer is freed after action is performed.
+    /// - Returns: return value of `action` closure
+    @discardableResult private func withSphereFS<T>(
+        sphereIdentity: String,
+        action: (OpaquePointer) throws -> T
+    ) throws -> T {
+        let sphereIdentityPointer: UnsafeMutablePointer<CChar> = sphereIdentity
+            .withCString { pointer in
+                UnsafeMutablePointer(mutating: pointer)
+            }
+        defer {
+            ns_string_free(sphereIdentityPointer)
+        }
+
+        guard let sphereFS = ns_sphere_fs_open(
+            noosphere,
+            sphereIdentityPointer
+        ) else {
+            throw NoosphereError.foreignError("Failed to get pointer for sphere file system")
+        }
+        defer {
+            ns_sphere_fs_free(sphereFS)
+        }
+        return try action(sphereFS)
+    }
+
+    /// Read the value of a memo from a Sphere
+    /// - Returns: `SphereMemo`
+    func read(
+        sphereIdentity: String,
+        path: String
+    ) throws -> SphereMemo {
+        try withSphereFS(sphereIdentity: sphereIdentity) { sphereFS in
+            guard let file = ns_sphere_fs_read(noosphere, sphereFS, path) else {
+                throw NoosphereError.memoDoesNotExist
+            }
+            defer {
+                ns_sphere_file_free(file)
+            }
+
+            let contentTypeValues = ns_sphere_file_header_values_read(
+                file,
+                "Content-Type"
+            )
+            defer {
+                ns_string_array_free(contentTypeValues)
+            }
+
+            let contentType = String(cString: contentTypeValues.ptr.pointee!)
+
+            let contents = ns_sphere_file_contents_read(noosphere, file)
+            defer {
+                ns_bytes_free(contents)
+            }
+
+            let data: Data = Data(bytes: contents.ptr, count: contents.len)
+
+            return SphereMemo(contentType: contentType, data: data)
+        }
+    }
+
+    func write(
+        sphereIdentity: String,
+        path: String,
+        contentType: String,
+        contents: Data
+    ) throws {
+        try withSphereFS(sphereIdentity: sphereIdentity) { sphereFS in
+            contents.withUnsafeBytes({ rawBufferPointer in
```

Comment:
Had to update withUnsafeBytes usage, since the form used in tests is now deprecated. New API gives you an UnsafeRawBufferPointer https://developer.apple.com/documentation/swift/array/withunsafebytes(_:)